File: WGBS/CpG_compress_to_bin.R
rm(list=ls())
.libPaths(c("/storage/zhangyanxiaoLab/suzhuojie/R/x86_64-pc-linux-gnu-library/4.2/"))
setwd("/storage/zhangyanxiaoLab/suzhuojie/projects/Aging_CUT_Tag/")
set.seed(1)
library(dplyr)
library(tidyr)
library(stringr)
library(data.table)  
library(parallel)  
args <- commandArgs(trailingOnly = TRUE)  
if (length(args) < 1) {  
  stop("No tissue argument provided")  
}  
tissue <- args[1]  
bin_size <- args[2]
print(paste("Tissue is:", tissue))  
print(paste("bin size is:", bin_size))  

dir.create(paste0("data/samples/WGBS/",tissue,"/compress2bin/"))
search_table <- read.csv("data/samples/all/WGBS_search_table.csv")


compress_to_bin <- function(tissue, bin_size){
  samples <- search_table$sample_name[which(search_table$tissue==tissue)]
  bin_summary <- data.frame(bin = character(),
                            total_V4 = numeric(),
                            total_V5 = numeric(),
                            percent = numeric(),
                            tissue = character(),
                            sample = character())
  for(sample in samples){
    df <- fread(paste0("data/samples/WGBS/",tissue,"/bdg/",sample,"_CpG.bdg"),sep = "\t")
    df <- df[which(df$V1 %in% paste0("chr",c(1:19,"X","Y"))),]
    df$V1 <- factor(df$V1,paste0("chr",c(1:19,"X","Y")))
    df$V3 <- df$V2
    df <- df[order(df$V1,df$V2),]
    
    ref <- fread(paste0("~/ref_data/mm10_", bin_size, "_bins.bed")) 
    ref$V1 <- factor(ref$V1,paste0("chr",c(1:19,"X","Y")))
    ref <- ref[order(ref$V1,ref$V2),]
    ref$V2 <- ref$V2+1
    ref[, `:=`(total_V4 = NA, total_V5 = NA)]  
    
    setDT(df)  
    setDT(ref)  
    setkey(df, V1, V2, V3) 
    setkey(ref, V1, V2, V3) 
    overlaps <- foverlaps(df, ref, type = "any", nomatch = 0L)  
    results <- overlaps[, .(total_V4 = sum(i.V4), total_V5 = sum(V5)), by = .(V4)]  
    results$percent <- results$total_V4/results$total_V5
    colnames(results)[1] <- "label"
    results$tissue <- tissue
    results$sample <- sample
    bin_summary <- rbind(bin_summary,results)
    # t_df <- df %>% filter(V2 > "27570001" & V2 <= "27580000" & V1 == "chr1") #check
  }
  write.csv(bin_summary,paste0("data/samples/WGBS/",tissue,"/compress2bin/",bin_size,"_bins_all_depth.csv"),row.names = F)
}
compress_to_bin(tissue,bin_size)

File: specific_modality/WGBS/CpG_compress_to_bin.py
import os  
import pandas as pd  
import numpy as np  
import multiprocessing as mp  
os.chdir("/storage/zhangyanxiaoLab/suzhuojie/projects/Aging_CUT_Tag/")  

def process_chunk(chunk, ref):  
    ref_sorted = ref.sort_values(by=['V1', 'V2']).reset_index(drop=True)  
    chunk_sorted = chunk.sort_values(by=['V1', 'V2']).reset_index(drop=True)  
    merged = pd.merge_asof(chunk_sorted, ref_sorted, by='V1', left_on='V2', right_on='V2', direction='backward')  
    merged = merged[(merged['V2_y'] <= merged['V2_x']) & (merged['V3'] >= merged['V2_x'])]  
    chunk['bin_id'] = merged['V4']  
    return chunk  

def compress_to_bin(tissue, bin_size):  
    search_table = pd.read_csv("data/samples/all/WGBS_search_table.csv")  
    samples = search_table[search_table['tissue'] == tissue]['sample_name']  
    bin_methylation_summary = pd.DataFrame(columns=['bin_id', 'total_V4', 'total_V5', 'percent', 'sample'])  
    os.makedirs(f"data/samples/WGBS/{tissue}/compress2bin/", exist_ok=True)  
    ref = pd.read_csv(f"~/ref_data/mm10_{bin_size}_bins.bed", sep="\t", header=None, names=['V1', 'V2', 'V3', 'V4'])  
    ref = ref.sort_values(by=['V1', 'V2']).reset_index(drop=True)  
    for sample in samples:  
        df = pd.read_csv(f"data/samples/WGBS/{tissue}/bdg/{sample}_CpG.bdg", sep="\t", header=None, names=['V1', 'V2', 'V3', 'V4', 'V5', 'V6']) 
        # df['depth'] = df['V5']  
        chromosomes = [f"chr{i}" for i in range(1, 20)] + ["chrX", "chrY"]
        df = df[df['V1'].isin(chromosomes)]  
        df = df.sort_values(by=['V1', 'V2']).reset_index(drop=True)  
        num_cores = 10  
        chunks = np.array_split(df, num_cores)  
        with mp.Pool(num_cores) as pool:  
            results = pool.starmap(process_chunk, [(chunk, ref) for chunk in chunks])  
        final_result = pd.concat(results, ignore_index=True)  
        final_result.to_csv(f"data/samples/WGBS/{tissue}/compress2bin/{sample}_CpG_compress_to_{bin_size}_bin_level_all_depth.csv", index=False)  
        bin_methylation = final_result.groupby('bin_id').agg({'V4': 'sum', 'V5': 'sum'}).reset_index()  
        bin_methylation['percent'] = bin_methylation['V4'] / bin_methylation['V5'] * 100  
        bin_methylation['sample'] = sample  
        bin_methylation_summary = pd.concat([bin_methylation_summary, bin_methylation], ignore_index=True)  
    
    bin_methylation_summary.to_csv(f"data/samples/WGBS/{tissue}/{bin_size}_bin_level_CpG_all_depth.csv", index=False)  
